rfcs/react-components/convergence/as-extensibility.md

```diff
@@ -0,0 +1,110 @@
+# RFC: Adding extensibility path to the `as` prop
+
+---
+
+@khmakoto
+
+## Summary
+
+We had previously restricted the `as` prop to accept only HTML elements due to issues this created with the developer experience. While this makes sense in general, we might be able to add a bit of an extensibility story while still supporting most of the restrictions we still have in place.
+
+## Proposal
+
+What we are proposing here is to allow slots typed as `slotName: Slot<typeof Component>` to be able to be passed in only components whose props extend the props interface of the original type passed in.
+
+A prime example of this would be having the `primaryActionButton` slot of the `SplitButton` be a `ToggleButton` instead of a regular button.
+
+`SplitButtonSlots` type definition:
+
+```ts
+export type SplitButtonSlots = {
+  /**
+   * Root of the component that wraps the primary action button and menu button.
+   */
+  root: NonNullable<Slot<'div'>>;
+
+  /**
+   * Button that opens menu with secondary actions in SplitButton.
+   */
+  menuButton?: Slot<typeof MenuButton>;
+  /**
+   * Button to perform primary action in SplitButton.
+   */
+  primaryActionButton?: Slot<typeof Button>;
+};
+```
+
+`ToggleButtonProps` type definition:
+
+```ts
+type ToggleButtonCommons = {
+  /**
+   * Defines the controlled checked state of the `ToggleButton`.
+   * If passed, `ToggleButton` ignores the `defaultChecked` property.
+   * This should only be used if the checked state is to be controlled at a higher level and there is a plan to pass the
+   * correct value based on handling `onClick` events and re-rendering.
+   * @default false
+   */
+  checked: boolean;
+};
+
+export type ToggleButtonProps = ButtonProps &
+  Partial<ToggleButtonCommons> & {
+    /**
+     * Defines whether the `ToggleButton` is initially in a checked state or not when rendered.
+     * @default false
+     */
+    defaultChecked?: boolean;
+  };
+```
+
+Since `primaryActionButton` is typed as `Slot<typeof Button>` and `ToggleButtonProps` extends from `ButtonProps`, the idea is that we could do:
+
+```tsx
+<SplitButton primaryActionButton={{ as: ToggleButton }} />
+```
+
+And it would work correctly replacing the primary action button with a `ToggleButton`, while at the same time prohibiting components whose props do not extend those of `Button`. For example:
+
+```tsx
+<SplitButton primaryActionButton={{ as: Menu }} />
```

Comment:
There is historical context missing here about why we do not support the `as` prop usage in this way.

```tsx
// renders an `a` tag for the `Button` component
<Button as="a">Link button</Button>

// no longer renders the `Button` component but a plain `a` tag
<ComponentWithButtonSlot button={{ as: "a" }}  />
```

The above two examples behave in different ways, and the `as` prop usage now has two different behaviours depending on whether it is used in a slot or as a direct JSX property.

#17464 was written explicitly detailing the problems of accepting react elements as values for `as`. This also includes an interesting performance problem related to event listeners.

The above RFC was how we settled on the API for recomposition of components for the scenario that you are desecribing

```ts
function SplitButton(props: SplitButtonProps) {
  const { state, render } = useSplitButton({
    components: { primaryActionButton: Menu },
  });

  return render(state);
}
```

Reply:
5 cents: It's also not possible to compose `as` prop:

```tsx
// "Button" can render "a" tag
<Button as="a" />
// "ComponentWithButtonSlot" renders "Button", i.e.
//   state.components.button = Button
<ComponentWithButtonSlot button={{ children: 'foo' }} />
// To render "CustomButton" as "a" we will need to do
// something like Reakit did 💣
<ComponentWithButtonSlot button={{ as: [CustomButton, 'a'] }} /> 
```

I don't think that the last example shows something that we would like support.

---

I think that there is a miss-understanding that comes from the special handling for slots. `as` prop is a way to change element type for a component, not component's type: 

```tsx
// What we are actually doing
React.createElement(Button, null, { as: 'a' })
// What proposes this RFC
React.createElement('a', null)
```

The fair note is that it works currently for slots in this way when there are primitives, i.e.:

```tsx
if (typeof props.as === 'string') {
  React.createElement(components.SLOT, null, { as: props.as })
} else {
  React.createElement(props.as, null)
}
```

If the behavior above is confusing I would rather consider changing it.